apps/web/features/admin/rate-limit/rate-limit-form.tsx

```diff
@@ -0,0 +1,97 @@
+"use client";
+
+import { zodResolver } from "@hookform/resolvers/zod";
+import { useForm } from "react-hook-form";
+import { z } from "zod";
+
+import {
+  Form,
+  FormControl,
+  FormField,
+  FormItem,
+  FormLabel,
+  FormMessage,
+} from "@repo/ui/components/form";
+import { Input } from "@repo/ui/components/input";
+import { toast } from "sonner";
+import { Button } from "@repo/ui/components/button";
+import { submitRateLimit } from "@/lib/actions/admin/rate-limit";
+import { useTransition } from "react";
+import type {
+  RateLimitResponse,
+  RateLimitResponseData,
+} from "@/types/rate-limit";
+import { TooltipWrapper } from "@/components/tooltip-wrapper";
+
+const FormSchema = z.object({
+  rate: z.coerce
+    .number()
+    .min(1, "Rate limit must not be less than 1")
+    .max(1000, "Rate limit must not be more than 1000"),
+});
+
+type Props = {
+  defaultValue: RateLimitResponseData | null;
+};
+
+export function RateLimitForm({ defaultValue }: Props) {
+  const [isPending, setTransition] = useTransition();
+  const form = useForm<z.infer<typeof FormSchema>>({
+    resolver: zodResolver(FormSchema),
+    defaultValues: {
+      rate: defaultValue?.limit || 0,
+    },
+  });
+
+  function onSubmit(data: z.infer<typeof FormSchema>) {
+    const rateLimit = data.rate;
+    setTransition(async () => {
```

Comment:
https://react.dev/reference/react/useTransition#starttransition-parameters

I think it's a misuse of `useTransition`. The doc says that the action expects to contain setState, but you have no in your action

Reply:
> A state update marked as a Transition will be interrupted by other state updates. For example, if you update a chart component inside a Transition, but then start typing into an input while the chart is in the middle of a re-render, React will restart the rendering work on the chart component after handling the input update.

You don't have such case here, so you can use common states to handle loading logic